[MKC-1191] Bluetooth® Low Energy Learn Article (#1367)
arduino/docs-content: content/learn/05.communication/08.bluetooth/bluetooth.md

---
title: 'Bluetooth® Low Energy'
description: 'Bluetooth® LE is a wireless communication technology designed for short-range communication between electronic devices.'
author: Hannes Siebeneicher
tags: [bluetooth]
difficulty: 'intermediate'
---

## Introduction

Bluetooth® is a wireless communication technology that enables data exchange between electronic devices over short distances. It emerged during the 90s as an alternative to data cables eliminating the need for a physical connection.

Today it’s often used when connecting a phone to a Bluetooth® speaker or transferring data between two devices. In those scenarios, Bluetooth is designed for relatively high data transfer rates and continuous communication. However, with the growing demand for energy-efficient wireless communication in applications such as fitness tracking, healthcare devices, and Internet of Things (IoT) devices, there was a need for a technology optimized for low power consumption.

> This resulted in Bluetooth Low Energy (BLE) which emerged as part of the Bluetooth 4.0 specification, officially released in 2010. This marked the introduction of BLE as a separate and distinct technology within the broader Bluetooth ecosystem.

Reviewer: We are not allowed to use the abbreviation "BLE". I would add a note saying that they are connected but then keep using the "Bluetooth LE" abbreviation.

> ***Note: Bluetooth® Low Energy is not compatible with Bluetooth® classic***

Reviewer: ```suggestion
***Note: Bluetooth Low Energy is not compatible with Bluetooth® classic.***
```

> This article focuses on BLE, how it works and how it can be used together with compatible Arduino boards.

Reviewer: ```suggestion
This article focuses on Bluetooth LE, how it works and how it can be used together with compatible Arduino boards.
```

### Harald Blåtand

> Harald Blåtand was a 10th-century Danish king who ruled from approximately 958 to 986 AD and he originally inspired the name Bluetooth®. He is best known for his role in uniting various Danish tribes and for his conversion to Christianity, which had a significant impact on the history of Denmark. His nickname “Blåtand” (translated: Bluetooth) is believed to have been inspired by his dead or discoloured tooth, which may have appeared blue or black.

Reviewer: ```suggestion
Harald Blåtand was a 10th-century Danish king who ruled from approximately 958 to 986 AD and he originally inspired the name Bluetooth®. He is best known for his role in uniting various Danish tribes and for his conversion to Christianity, which had a significant impact on the history of Denmark. His nickname “Blåtand” (translated: Bluetooth) is believed to have been inspired by his dead or discolored tooth, which may have appeared blue or black.
```

Similar to how Harald Blåtand united Denmark the Bluetooth protocol was meant to unite various devices and communication protocols.

## Technical Specifications

The following parts explore the core concepts and technical specifications of Bluetooth® LE.

### Frequency Bands and Range

> Bluetooth® Low Energy (BLE) operates in the 2.4 GHz ISM (Industrial, Scientific, and Medical) band, which is commonly used for various wireless technologies. This frequency band is divided into multiple channels that BLE devices use for communication.

Reviewer: ```suggestion
Bluetooth® Low Energy (Bluetooth LE) operates in the 2.4 GHz ISM (Industrial, Scientific, and Medical) band, which is commonly used for various wireless technologies. This frequency band is divided into multiple channels that Bluetooth LE devices use for communication.
```

> ### Typical BLE Range

Reviewer: ```suggestion
### Typical Bluetooth LE Range
```

The range of a Bluetooth® LE connection can vary depending on several factors, but in typical scenarios, it can extend up to approximately 50 meters (or roughly 164 feet) in a line-of-sight environment. This range can be affected by several factors:

> **Obstacles**: Physical obstacles such as walls, furniture, and other objects can significantly reduce the range of a BLE connection. Thick walls and materials like concrete can be particularly challenging for Bluetooth® LE signals to penetrate.

Reviewer: ```suggestion
**Obstacles**: Physical obstacles such as walls, furniture, and other objects can significantly reduce the range of a connection. Thick walls and materials like concrete can be particularly challenging for Bluetooth® LE signals to penetrate.
```

> **Interference**: As mentioned earlier, the 2.4 GHz band is shared by various wireless devices. Interference from other devices operating in the same frequency range can impact the range and reliability of BLE connections.

Reviewer: ```suggestion
**Interference**: As mentioned earlier, the 2.4 GHz band is shared by various wireless devices. Interference from other devices operating in the same frequency range can impact the range and reliability of Bluetooth LE connections.
```

**Antenna Design**: The design and quality of the antennas in both the central and peripheral devices can influence the range. Devices with well-designed antennas tend to have better coverage.

**Orientation**: The relative orientation of the central and peripheral devices also affects range. A clear line of sight between devices typically results in the best range, while obstructed lines of sight can reduce it.

### Central / Peripheral Devices

> Bluetooth® Low Energy devices operate using different roles and modes that define how devices interact with eacher other.

Reviewer: ```suggestion
Bluetooth® Low Energy devices operate using different roles and modes that define how devices interact with each other.
```

**Central Device**: A central device in Bluetooth®  LE is typically a more capable device with features like a higher CPU power, more memory, or a larger battery. Central devices take on the role of initiating connections to peripheral devices. For example, your smartphone is often a central device when connecting to BLE peripherals like fitness trackers or smart sensors.

> **Peripheral Device**: Peripheral devices are generally resource-constrained compared to central devices. They advertise their presence and data to central devices. Compared to Bluetooth® classic Bluetooth® LE devices don't maintain a continues connection to the central device to save power. Examples of peripheral devices include environmental sensors or health monitors.

Reviewer: ```suggestion
**Peripheral Device**: Peripheral devices are generally resource-constrained compared to central devices. They advertise their presence and data to central devices. Compared to Bluetooth® classic, Bluetooth® LE devices don't maintain a continues connection to the central device to save power. Examples of peripheral devices include environmental sensors or health monitors.
```

> ![BLE Roles](./assets/ble_roles.png)

Reviewer: ```suggestion
![Bluetooth LE Roles](./assets/ble_roles.png)
```

### Advertising / Connection Mode

**Advertising Mode**: Advertising mode is primarily used to make a Bluetooth® LE peripheral device discoverable by other devices, particularly central devices. During advertising mode, the peripheral device periodically broadcasts advertising packets. These packets contain information about the peripheral's identity, services, and characteristics. Central devices continuously scan for these advertising packets to discover nearby peripherals. In advertising mode, the peripheral device is not actively connected to any central device. It remains in a low-power state while broadcasting advertising packets. It is "waiting" for a central device to establish a connection.

> **Connection Mode**: Connection mode is activated once a central device successfully establishes a connection with a peripheral device. During this mode, devices can exchange data bi-directionally. Central devices can read data from and write data to the peripheral device. The connection mode is crucial for ongoing communication between BLE devices.

Reviewer: ```suggestion
**Connection Mode**: Connection mode is activated once a central device successfully establishes a connection with a peripheral device. During this mode, devices can exchange data bi-directionally. Central devices can read data from and write data to the peripheral device. The connection mode is crucial for ongoing communication between Bluetooth LE devices.
```

> ![BLE Modes](./assets/ble_mode.gif)

Reviewer: ```suggestion
![Bluetooth LE Modes](./assets/ble_mode.gif)
```

**Defining Services:**

In Bluetooth® LE, a service can be thought of as a logical grouping of related data measurements or functionalities provided by a peripheral device. These data measurements can represent various aspects of the device's capabilities or the information it collects.

For example, consider a weather monitoring sensor. It might have a service called "Weather Data" that encompasses measurements like temperature, humidity, and wind speed. Another service, "Energy Information," could include data related to battery level and energy consumption.

**Introducing Characteristics**:
Within each service, we have characteristics. Characteristics are individual data points or attributes that provide specific information or measurements.

For instance, in the "Weather Data" service mentioned earlier, characteristics may include "Temperature," "Humidity," and "Wind Speed." These characteristics continuously record data and update as new measurements become available.

Similarly, the "Energy Information" service may consist of characteristics like "Battery Level" and "Energy Consumption."

> **UUIDs (Unique Universal Identifier)**

Reviewer: ```suggestion
**UUIDs (Unique Universal Identifier):**
```

To distinguish services and characteristics, Bluetooth® LE relies on a unique identifier called a UUID (Unique Universal Identifier).

A UUID is a 128-bit value that serves as a universally unique name for a service or characteristic. It acts like a label or identifier that central devices use to identify and communicate with specific services and characteristics.

UUIDs play a crucial role in Bluetooth® LE communication because they ensure that central devices can accurately locate and interact with the desired data points on peripheral devices. They eliminate ambiguity and allow for precise data retrieval and control.

> In practical terms, understanding services and characteristics is essential when designing or interacting with BLE devices. Services provide a high-level organization of data, while characteristics represent the individual data points within those services. UUIDs act as the keys that enable central devices to access and utilize the data provided by peripheral devices.

Reviewer: ```suggestion
In practical terms, understanding services and characteristics is essential when designing or interacting with Bluetooth LE devices. Services provide a high-level organization of data, while characteristics represent the individual data points within those services. UUIDs act as the keys that enable central devices to access and utilize the data provided by peripheral devices.
```

> As you explore Bluetooth® LE further, you'll encounter various predefined services and characteristics used in common applications. These standardized profiles simplify the development process, making it easier to create BLE-based projects and applications.

Reviewer: ```suggestion
As you explore Bluetooth® LE further, you'll encounter various predefined services and characteristics used in common applications. These standardized profiles simplify the development process, making it easier to create Bluetooth LE-based projects and applications.
```

Reviewer: @Hannes7eicher you get the point, update all BLE to Bluetooth LE 😁

**Profiles**

Bluetooth® LE profiles are predefined sets of services and characteristics that standardize how BLE devices interact with each other. These profiles define the behavior and capabilities of BLE devices, making it easier for different devices to communicate seamlessly. Let's delve into Bluetooth® LE profiles:

**Defining BLE Profiles**:

Bluetooth® LE profiles serve as blueprints that specify how data should be organized and exchanged between devices in a standardized manner. They define the roles, services, and characteristics that devices can use to communicate effectively.

Each profile is tailored to a specific use case or application, ensuring that devices of different manufacturers can work together seamlessly when using the same profile.

> **Common Standard Profiles**:

Reviewer: ```suggestion
**Common Standard Profiles:**
```

Bluetooth® LE includes a range of standard profiles that simplify the development of BLE applications. Some of the most well-known standard profiles include:

- Battery Service: The Battery Service provides information about the battery level of a device. It typically includes a Battery Level characteristic that central devices can read to monitor the battery status of a peripheral device, such as a wireless headset or smartwatch.

- Heart Rate Service: The Heart Rate Service is commonly used in fitness and health monitoring applications. It includes characteristics that provide real-time heart rate data, allowing central devices like smartphones or fitness trackers to monitor a user's heart rate during exercise.

- Generic Access Profile (GAP): While not a service in itself, GAP defines the roles and procedures for device discovery and connection establishment in BLE. It plays a vital role in enabling devices to find and connect to each other seamlessly.

> ***You can read more about Bluetooth® profiles [here](https://en.wikipedia.org/wiki/List_of_Bluetooth_profiles)***

Reviewer: ```suggestion
***You can read more about Bluetooth® profiles [here](https://en.wikipedia.org/wiki/List_of_Bluetooth_profiles).***
```

> **Creating Custom Profiles** :

Reviewer: ```suggestion
**Creating Custom Profiles:**
```

Author: I removed them.

In addition to standard profiles, developers have the flexibility to create custom profiles tailored to their specific application needs. These custom profiles define unique services and characteristics that match the requirements of a particular project.

By using Bluetooth® LE profiles, developers can leverage existing standardized profiles for common applications or create custom profiles for specialized projects. This standardized approach simplifies the development process, enhances interoperability, and allows for the creation of diverse BLE-based applications, from health monitoring to home automation.

As you explore Bluetooth® LE further, you'll discover a wide range of profiles designed to support various use cases. These profiles play a crucial role in ensuring that BLE devices can seamlessly communicate and provide valuable data to central devices.

## ArduinoBLE Library

> The next part introduces how do establish a Bluetooth® LE connectionn, turn on the built in LED using your phone and read analog values via Bluetooth® LE. We will go through the following steps in order to create our sketch:

Reviewer: ```suggestion
The next part introduces how do establish a Bluetooth® LE connection, turn on the built in LED using your phone and read analog values via Bluetooth® LE. We will go through the following steps in order to create our sketch:
```

- Create a new service.
- Create an analog pin characteristic.
- Set the name for our device.
- Start advertising the device.
- Create a conditional that works only if an external device is connected (smartphone).
- Create a conditional that turns on an LED over Bluetooth®.

> - Read an analog pin over Bluetooth®.

Reviewer: Should this be a numbered list instead?

> First, let's make sure we have the drivers for the board you are using installed. If we are using the Web Editor, we do not need to install anything. If we are using an offline editor, we need to install it manually. This can be done by navigating to **Tools > Board > Board Manager....**

Reviewer: ```suggestion
First, let's make sure we have the drivers for the board you are using installed. If we are using the Arduino Web Editor, we do not need to install anything. If we are using an offline editor, we need to install it manually. This can be done by navigating to **Tools > Board > Board Manager....**.
```

Now, we need to install the library needed. If we are using the Web Editor, there is no need to install anything. If we are using an offline editor, simply go to **Tools > Manage libraries..**, and search for ArduinoBLE and install it.

![Install ArduinoBLE](./assets/install_library.png)

___

> >**NOTE:** This section is optional, you can find the complete code further down on this tutorial.

Reviewer: If we are giving the reader an option to skip the section, include either a link to the next section or tell them where they can go to skip this part.

Author: fixed

First, we need to include the **ArduinoBLE** library, and create a new service. We will name the service **"180A"** which is translated to **"Device Information"**.  We will then create two characteristics, one for the LED, and one for the analog pin. The name **"2A57"** is translated to **"Digital Output"** and **"2A58** is translated to **"Analog"**. 

```arduino
#include <ArduinoBLE.h
BLEService newService("180A"); // creating the service

BLEUnsignedCharCharacteristic randomReading("2A58", BLERead | BLENotify); // creating the Analog Value characteristic
BLEByteCharacteristic switchChar("2A57", BLERead | BLEWrite); // creating the LED characteristic

long previousMillis = 0;

```

> In the `setup()`, we will start by initializing serial communication, define both the in-built LED and the LED we connected to pin 2, and initialize the **ArduinoBLE** library.

Reviewer 2: ```suggestion
In the `setup()`, we will start by initializing serial communication, define both the built-in LED and the LED we connected to pin 2, and initialize the **ArduinoBLE** library.
```

We then set the name for our device, using the command `BLE.setLocalName("Arduino Board");`, then add the characteristics we defined previously to the service created earlier, `newService`. After they have been added, we will also add the service, using the command `BLE.addService(newService);`.

The final steps we will take is to set the starting value of 0 for both characteristics. This is mostly important for the LED, since 0 means it will be OFF from the start.

The setup is finished by using the command `BLE.advertise();`, which makes it visible for other devices to connect to.

```arduino
void setup() {
  Serial.begin(9600);    // initialize serial communication
  while (!Serial);       //starts the program if we open the serial monitor.

  pinMode(LED_BUILTIN, OUTPUT); // initialize the built-in LED pin to indicate when a central is connected
  pinMode(ledPin, OUTPUT); // initialize the built-in LED pin to indicate when a central is connected

  //initialize ArduinoBLE library
  if (!BLE.begin()) {
    Serial.println("starting Bluetooth® Low Energy failed!");
    while (1);
  }

>   BLE.setLocalName("Arduino Board"); //Setting a name that will appear when scanning for Bluetooth® devices

Reviewer: ```suggestion
  BLE.setLocalName("Arduino Board"); //Setting a name that will appear when scanning for Bluetooth devices
```

  newService.addCharacteristic(switchChar); //add characteristics to a service
  newService.addCharacteristic(randomReading);

  BLE.addService(newService);  // adding the service

  switchChar.writeValue(0); //set initial value for characteristics
  randomReading.writeValue(0);

  BLE.advertise(); //start advertising the service
  Serial.println(" Bluetooth® device active, waiting for connections...");
}
```

In the `loop()` we will use the command `BLEDevice central = BLE.central();` to start waiting for a connection. When a device connects, the address of the connecting device (the central) will be printed in the Serial Monitor, and the in-built LED will turn ON. 

After this, we use a `while` loop that only runs as long as a device is connected. Here, we do a reading of Analog pin 1, which will record random values between 0 and 1023. We then use a conditional to check if there's an incoming value: if any value other than 0 comes in, the LED turns ON, but if 0 comes in, it turns it OFF.

If our device (smartphone) disconnects, we exit the `while` loop. Once it exits, it prints the message **"Disconnected from central"** in the Serial Monitor. 

```cpp
void loop() {
  

>   BLEDevice central = BLE.central(); // wait for a Bluetooth® Low Energy central

Reviewer: ```suggestion
  BLEDevice central = BLE.central(); // wait for a Bluetooth Low Energy central
```

  if (central) {  // if a central is connected to the peripheral
    Serial.print("Connected to central: ");
    

>     Serial.println(central.address()); // print the central's BT address

Reviewer: ```suggestion
    Serial.println(central.address()); // print the central's Bluetooth address
```

    
    digitalWrite(LED_BUILTIN, HIGH); // turn on the LED to indicate the connection

    
    while (central.connected()) { // while the central is connected:
      long currentMillis = millis();
      
      if (currentMillis - previousMillis >= 200) { 
        previousMillis = currentMillis;

        int randomValue = analogRead(A1);
        randomReading.writeValue(randomValue);

        if (switchChar.written()) {
          if (switchChar.value()) {   // any value other than 0
            Serial.println("LED on");
            digitalWrite(ledPin, HIGH);         // will turn the LED on
          } else {                              // a 0 value
            Serial.println(F("LED off"));
            digitalWrite(ledPin, LOW);          // will turn the LED off
          }
        }

      }
    }
    
    digitalWrite(LED_BUILTIN, LOW); // when the central disconnects, turn off the LED
    Serial.print("Disconnected from central: ");
    Serial.println(central.address());
  }
}
```

## Complete Code

> If you choose to skip the code building section, the complete code can be found below:

Reviewer: ```suggestion
If you chose to skip the code building section, the complete code can be found below:
```

```arduino
#include <ArduinoBLE.h>
BLEService newService("180A"); // creating the service

BLEUnsignedCharCharacteristic randomReading("2A58", BLERead | BLENotify); // creating the Analog Value characteristic
BLEByteCharacteristic switchChar("2A57", BLERead | BLEWrite); // creating the LED characteristic

long previousMillis = 0;

void setup() {
  Serial.begin(9600);    // initialize serial communication
  while (!Serial);       //starts the program if we open the serial monitor.

  pinMode(LED_BUILTIN, OUTPUT); // initialize the built-in LED pin to indicate when a central is connected

  //initialize ArduinoBLE library
  if (!BLE.begin()) {
    Serial.println("starting Bluetooth® Low Energy failed!");
    while (1);
  }

>   BLE.setLocalName("Arduino Board"); //Setting a name that will appear when scanning for Bluetooth® devices

Reviewer: ```suggestion
  BLE.setLocalName("Arduino Board"); //Setting a name that will appear when scanning for Bluetooth devices
```

  BLE.setAdvertisedService(newService);

  newService.addCharacteristic(switchChar); //add characteristics to a service
  newService.addCharacteristic(randomReading);

  BLE.addService(newService);  // adding the service

  switchChar.writeValue(0); //set initial value for characteristics
  randomReading.writeValue(0);

  BLE.advertise(); //start advertising the service

>   Serial.println(" Bluetooth® device active, waiting for connections...");

Reviewer: You can remove the trademark symbol from here

      
      if (currentMillis - previousMillis >= 200) { // if 200ms have passed, we check the battery level
        previousMillis = currentMillis;

        int randomValue = analogRead(A1);
        randomReading.writeValue(randomValue);

        if (switchChar.written()) {
          if (switchChar.value()) {   // any value other than 0
            Serial.println("LED on");
            digitalWrite(LED_BUILTIN, HIGH);         // will turn the LED on
          } else {                              // a 0 value
            Serial.println(F("LED off"));
            digitalWrite(LED_BUILTIN, LOW);          // will turn the LED off
          }
        }

      }
    }
    
    digitalWrite(LED_BUILTIN, LOW); // when the central disconnects, turn off the LED
    Serial.print("Disconnected from central: ");
    Serial.println(central.address());
  }
}
```

## Testing It Out

Once we are finished with the coding, we can upload the sketch to the board. When it has been successfully uploaded, open the Serial Monitor. In the Serial Monitor, the text **" Bluetooth® device active, waiting for connections..."** will appear. 

![Waiting for connections.](./assets/waiting_connection.png)

> We can now discover our MKR WiFi 1010 board in the list of available Bluetooth® devices. To access the service and characteristic we recommend using the **LightBlue** application. Follow <a href="https://apps.apple.com/us/app/lightblue/id557428110">this link for iPhones</a> or <a href="https://play.google.com/store/apps/details?id=com.punchthrough.lightblueexplorer&hl=en">this link for Android phones</a>. 

Reviewer 2: ```suggestion
We can now discover our Arduino MKR WiFi 1010  in the list of available Bluetooth® devices. To access the service and characteristic we recommend using the **LightBlue** application. Follow <a href="https://apps.apple.com/us/app/lightblue/id557428110">this link for iPhones</a> or <a href="https://play.google.com/store/apps/details?id=com.punchthrough.lightblueexplorer&hl=en">this link for Android phones</a>. 
```

- `BLE.stopScan()` - Stop scanning

- `BLE.disconnect()` - Disconnect from a peripheral

### Peripheral

- `BLE.setLocalName(name)` - Give your device a name

- `BLEService newService(service)` - Create a service object

> Read more about standard services in the [Assigned Numbers document](https://www.bluetooth.com/specifications/assigned-numbers/).

- `BLEUnsignedCharCharacteristic customCharacteristic("UUID_here", Properties_here);` - Create characteristic object

- `BLE.setAdvertisedService(bleService)` - Set the advertised service

- `bleService.addCharacteristic(bleCharacteristic)` - Add characteristics to the service

- `bleCharacteristic.writeValue()` - Write the value of the characteristic

- `BLE.advertise()` - Start advertising

***These are just a few of the most important methods. You can find more information about all methods and their details in the [ArduinoBLE reference](https://www.arduino.cc/reference/en/libraries/arduinoble/).***

## Examples

Below you can find examples showing how to send data between two Arduino boards and how to connect to your Arduino board, reading and writing values using your smartphone.

> ### Send data between two Arduino boards

Reviewer 3: ```suggestion
### Remote LED Control
```

> 

Reviewer 3: ```suggestion

This example can be used with two Bluetooth® supported Arduino boards, one with a button, and the other with an LED. When the button is pressed, it will advertise a corresponding value (high/low), which will be received by the board with an LED.

```

Reviewer 3: @Hannes7eicher just after this section, can we add a circuit diagram of the two board's setup? Basically the ones from the other tutorials, but compressed into one image if possible, like this:

```
#### Circuit

The following circuit is required for this example. Connect a button to one board (pin X), and an LED to the other (pin Y).

![Remote LED control circuit.](assets/image.png)

```

Author: Good idea!

> **Central**

Reviewer 3: ```suggestion
#### Central
```

    peripheral.disconnect();
    return;
  }

  // retrieve the LED characteristic
  BLECharacteristic ledCharacteristic = peripheral.characteristic("19b10001-e8f2-537e-4f6c-d104768a1214");

  if (!ledCharacteristic) {
    Serial.println("Peripheral does not have LED characteristic!");
    peripheral.disconnect();
    return;
  } else if (!ledCharacteristic.canWrite()) {
    Serial.println("Peripheral does not have a writable LED characteristic!");
    peripheral.disconnect();
    return;
  }

  while (peripheral.connected()) {
    // while the peripheral is connected

    // read the button pin
    int buttonState = digitalRead(buttonPin);

    if (oldButtonState != buttonState) {
      // button changed
      oldButtonState = buttonState;

      if (buttonState) {
        Serial.println("button pressed");

        // button is pressed, write 0x01 to turn the LED on
        ledCharacteristic.writeValue((byte)0x01);
      } else {
        Serial.println("button released");

        // button is released, write 0x00 to turn the LED off
        ledCharacteristic.writeValue((byte)0x00);
      }
    }
  }

  Serial.println("Peripheral disconnected");
}
```

> **Peripheral**

Reviewer 3: ```suggestion
#### Peripheral
```

  switchCharacteristic.writeValue(0);

  // start advertising
  BLE.advertise();

  Serial.println("BLE LED Peripheral");
}

void loop() {
  // listen for Bluetooth® Low Energy peripherals to connect:
  BLEDevice central = BLE.central();

  // if a central is connected to peripheral:
  if (central) {
    Serial.print("Connected to central: ");
    // print the central's MAC address:
    Serial.println(central.address());

    // while the central is still connected to peripheral:
    while (central.connected()) {
      // if the remote device wrote to the characteristic,
      // use the value to control the LED:
      if (switchCharacteristic.written()) {
        if (switchCharacteristic.value()) {   // any value other than 0
          Serial.println("LED on");
          digitalWrite(ledPin, HIGH);         // will turn the LED on
        } else {                              // a 0 value
          Serial.println(F("LED off"));
          digitalWrite(ledPin, LOW);          // will turn the LED off
        }
      }
    }

    // when the central disconnects, print it out:
    Serial.print(F("Disconnected from central: "));
    Serial.println(central.address());
  }
}
```

> ### Control you Arduino with a Smartphone

Reviewer 3: ```suggestion
### Control an Arduino with a Smartphone App
```

**Step 1 - Upload Code**
```arduino
#include <ArduinoBLE.h>
BLEService newService("180A"); // creating the service

BLEUnsignedCharCharacteristic randomReading("2A58", BLERead | BLENotify); // creating the Analog Value characteristic
BLEByteCharacteristic switchChar("2A57", BLERead | BLEWrite); // creating the LED characteristic

const int ledPin = 2;
long previousMillis = 0;

void setup() {
  Serial.begin(9600);    // initialize serial communication
  while (!Serial);       //starts the program if we open the serial monitor.

  pinMode(LED_BUILTIN, OUTPUT); // initialize the built-in LED pin to indicate when a central is connected
  pinMode(ledPin, OUTPUT); // initialize the built-in LED pin to indicate when a central is connected

  //initialize ArduinoBLE library
  if (!BLE.begin()) {
    Serial.println("starting Bluetooth® Low Energy failed!");
    while (1);
  }

>   BLE.setLocalName("MKR WiFi 1010"); //Setting a name that will appear when scanning for Bluetooth® devices

Reviewer 3: ```suggestion
  BLE.setLocalName("<My Board Name>"); //Setting a name that will appear when scanning for Bluetooth® devices
```

**Obstacles**: Physical obstacles such as walls, furniture, and other objects can significantly reduce the range of a Bluetooth LE connection. Thick walls and materials like concrete can be particularly challenging for Bluetooth LE signals to penetrate.

**Interference**: As mentioned earlier, the 2.4 GHz band is shared by various wireless devices. Interference from other devices operating in the same frequency range can impact the range and reliability of Bluetooth LE connections.

**Antenna Design**: The design and quality of the antennas in both the central and peripheral devices can influence the range. Devices with well-designed antennas tend to have better coverage.

**Orientation**: The relative orientation of the central and peripheral devices also affects range. A clear line of sight between devices typically results in the best range, while obstructed lines of sight can reduce it.

### Central / Peripheral Devices

Bluetooth Low Energy devices operate using different roles and modes that define how devices interact with each other.

> **Central Device**: A central device in Bluetooth  LE is typically a more capable device with features like a higher CPU power, more memory, or a larger battery. Central devices take on the role of initiating connections to peripheral devices. For example, your smartphone is often a central device when connecting to Bluetooth LE peripherals like fitness trackersy, smart sensors or an Arduino board.

Reviewer 4: ```suggestion
**Central Device**: A central device in Bluetooth  LE is typically a more capable device with features like a higher CPU power, more memory, or a larger battery. Central devices take on the role of initiating connections to peripheral devices. For example, your smartphone is often a central device when connecting to Bluetooth LE peripherals like fitness trackers, smart sensors or an Arduino board.
```

> **Peripheral Device**: Peripheral devices are generally resource-constrained compared to central devices (e.g. an Arduino board compared with your smartphone). Peripheral devices advertise their presence and data to central devices. Compared to Bluetooth classic, Bluetooth LE devices don't maintain a continues connection to the central device to save power.

Reviewer 4: ```suggestion
**Peripheral Device**: Peripheral devices are generally resource-constrained compared to central devices (e.g. an Arduino board compared with your smartphone). Peripheral devices advertise their presence and data to central devices. Compared to Bluetooth classic, Bluetooth LE devices don't maintain a continuous. connection to the central device to save power.
```

***You can read more about Bluetooth profiles [here](https://en.wikipedia.org/wiki/List_of_Bluetooth_profiles).***

**Creating Custom Profiles**

In addition to standard profiles, developers have the flexibility to create custom profiles tailored to their specific application needs. These custom profiles define unique services and characteristics that match the requirements of a particular project.

Using Bluetooth LE profiles, developers can leverage standardized profiles for common applications or create custom profiles for specialized projects. This standardized approach simplifies the development process, enhances interoperability, and allows for the creation of diverse Bluetooth LE-based applications, from health monitoring to home automation.

As you explore Bluetooth LE further, you'll discover a wide range of profiles designed to support various use cases. These profiles play a crucial role in ensuring that Bluetooth LE devices can seamlessly communicate and provide valuable data to central devices.

### Bluetooth Classic

> Bluetooth Low Energy is distinctively different from Bluetooth Classic. Bluetooth Classic operates in a manner similar to a serial port or UART (Universal Asynchronous Receiver-Transmitter), which is commonly used for point-to-point communication.

Reviewer 4: ```suggestion
Bluetooth Low Energy is distinctly different from Bluetooth Classic. Bluetooth Classic operates in a manner similar to a serial port or UART (Universal Asynchronous Receiver-Transmitter), which is commonly used for point-to-point communication.
```

-	Bluetooth Low Energy: sacrifices data transfer speed in favor of energy efficiency. It's ideal for applications that require intermittent or small bursts of data, such as sending sensor readings or control commands.

**Connection Types**

-	Bluetooth Classic: establishes a continuous and relatively power-hungry connection, making it suitable for applications requiring real-time, continuous communication.

-	Bluetooth Low Energy: supports two primary modes - advertising and connection. In advertising mode, a BLE peripheral periodically broadcasts its presence but doesn't maintain a continuous connection, conserving power. When needed, a central device can establish a connection for data exchange.

## ArduinoBLE Library

The [ArduinoBLE Library](https://www.arduino.cc/reference/en/libraries/arduinoble/) is the main library enabling Bluetooth Low Energy on compatible Arduino boards. You must first download and install the ArduinoBLE library. See our [instructions](https://docs.arduino.cc/software/ide-v2/tutorials/ide-v2-installing-a-library) on how to install a library. 

> In the following section you will find an overview and explanation of the librarie's most important methods and how to use them:

Reviewer 4: ```suggestion
In the following section you will find an overview and explanation of the library's most important methods and how to use them:
```
